### DISCUSSION TITLE
Modbus TCP server support

### DISCUSSION THREAD

**Discussion by mheyse** [emsesp/EMS-ESP32#1607] [Ideas]
Hi, I'm very new to the EMS-ESP project (just integrating it into my setup), so please forgive me if I missed something obvious.

This is my idea/feature request: Implement a Modbus server that exposes the EMS bus, in analogy to the existing MQTT and REST interfaces.

This would make interfacing with KNX components more robust because there are out-of-the-box KNX/Modbus gateways (e.g. the Weinzierl "KNX Modbus TCP Gateway 716"). I do have a server that also functions as a  MQTT <-> KNX gateway that does the job, but for "crtitical infrastructure" like heating/cooling  I prefer not having to rely on a PC/Raspberry/X1 etc.

Thanks for your thoughts!

Best
Michael

**Comment by proddy** [comment]
We've been thinking for a while about how to do the Modbus TCP (similar to https://husdata.se/docs/h60-manual/modbus-tcp-integrations) and have some of the code already done. It's just difficult to test without something like a Loxone. @bbqkees knows more...

**Reply by MichaelDvP** [reply]
There is a nice library https://github.com/eModbus/eModbus for the communication.
I think most work is to list and define the register-number for each entity. 
Maybe as entry in local_translations.h, next to mqtt shortname, than we know it's unique and does not change for new devices/entires?

**Reply by mheyse** [reply]
Thanks for your replies!

@MichaelDvP I think there is more to it then just defining register addresses. As a Modbus register can only hold a 16 bit integer value, a data type and encoding as well as the numbers of registers for the value need to be defined for each entity. Also, the register type (input/holding) must be defined depending on if it is read-only or read-write. There is no standard for the encoding, but most manufacturers use for example fixed-point numbers instead of floats, defining a division factor for each register. You also need to map enum values to integers.

For an example you could have a look at how SMA does that with their PV inverters here: https://files.sma.de/downloads/EDMx-Modbus-TI-en-16.pdf (data type and format definitions are on page 10, data point type definitions start on page 14).

I'd love to help with the implementation, but I'm not sure I'll find the time to work myself into the ems-esp source code, but I can offer to implement a Modbus client ("master" in the old terminology) in .net to test the Modbus interface.

**Reply by mheyse** [reply]
Hi @MichaelDvP @proddy,
I started implementing the modbus server using the eModbus lib which seems to work fine. I have some questions regarding the project, hope it is OK to ask here!

1) Modbus data value encoding: All entities already have a `DeviceValueNumOp` divisor/multiplier so I'll transmit the "raw" integer value via modbus and the client needs to know about the `numeric_operator` for each value to calculate the actual value. So that's already clear.

2) My plan to assign registers (now and for new entities in the future):
 - I would modify the generated `dump_entities.csv` and add columns with the the modbus start register (default to -1 if not yet defined) and the number of registers (default 0) in `EMSdevice::dump_value_info()` 
 - extract all entities using `dump_entities.sh`
 - add registers for all entities that don't have already registers assigned
 - calculate the number of registers from `DeviceValue::type`

3) Problem with this approach: `string` values. How do I get/define their maximum length? It seems they are currently hardcoded with fixed lengths in the individual devices (e.g. `Boiler::maintenanceDate_` is defined as `char[12]`) and also in the corresponding `set_...` methods. A register range needs to be reserved for each string value, so the maximum string length is also needed in the code that assigns the registers. I don't really want to manually extract the field lengths for all (282 in my dump here) string entities if possible.

4) Where to define the register and number of registers? I could put both values in `locale_translations.h` as you suggested, then the translations would start at offset 3 instead of 1, but I'm not sure this is great. Another option would be to have a mapping from the shortname to modbus parameters (register offset and length) in the modbus code. That would also make assigning registers with an automatic script easier, it could directly generate the source file with all registers and this would be independent from the `locale_translations.h` file. What do you think?

5) I would modify the entity dump and add columns with the `DeviceValue::numeric_operator`, the modbus start register and the number of registers in `EMSdevice::dump_value_info()` to be able to create documentation for modbus clients.

6) eModbus doesn't compile in native builds (mostly due to data type size mismatch) so I'll have to disable it completely in EMSESP_STANDALONE builds.

7) You don't seem to use the `lib_deps` platform.io-feature, but instead put all dependencies into the `lib` folder. So I'll add the eModbus source code to the `lib` folder, correct?

Thank you!
Michael

**Reply by MichaelDvP** [reply]
1. The numeric operator is an enum, but also contains the value of divider(positive) or multiplier(negative). Some operators are not easy to understand, like time in count-units of 10 or 15 minutes.
2. dump_entities goes through the list of device-library, the entities are multple times listed. On the other hand there are some devices that could have different functions, depending on device-id, for mixers this is in the csv with MM100/MM200, but the different solar functions are not included. Also remember that thermostats/mixers could have many heating circuits, hc1/seltemp is not the same as hc2/seltemp.
3. It's much less strings because of multiple entries. The longest is the last_error_code with 55 chars. There is a repeated discussion how to implements switchtimes and holiday modes, this could give longer strings in future.
4. Both possible. Using local_translations is difficult to handle different hcs. Maybe this is easier with extra list.
5. Good, this gives also the check for missing entries if using a second list for the registers.

**Reply by mheyse** [reply]
Please elaborate on the consequences of (2). In the CSV there is for example for both MM100 and MM200 an entity with shortname `pumpstatus` - both are booleans but have different fullnames, and one is writeable and one is not. So these should be different entities even though they have the same shortname. So I assume that different devices even of the same type have entirely different entity namespaces, so they are completely independent?

I think the cleanest solution (that also represents reality) is to implement multiple Modbus servers  - one for each device. eModbus easily supports this without overhead, and the `product id` could be equal to the Modbus `serverId`. And then there is a register/entity mapping that's individual for each device. This has the consequence that for example the `pumpstatus` entity might have different modbus addresses depending on the actual device.

For the different circuits I'd reserve blocks of registers, so for example 8 blocks for hc1...hc8. So each device would first have one register block with common values, followed by one register block for each circuit. I'd reserve some registers in each block for future extensions.

I'd also implement a Modbus server for ems-esp itself returning general information, so essentially the same information that is in the "info" mqtt topic, plus a list of all existing devices and also the installed circuits per device. Can I use ID `1` for this virtual device? This does not seem to be a product id used on the EMS bus.

**Reply by mheyse** [reply]
I just realized that the different `pumpstatus` entities are for different tags ("hc*" vs "wwc*") - so entity namespaces should probably be by device type and tag and don't need to be by device. This would imply that e.g. all devices of type "mixer" do have the same set of modbus registers, but some might not be used (for example, the MM100 would not use the "hc*" registers, only the "wwc*" and "dhw" registers).

[edit]: Some entity properties are different for different devices of the same type - most notably the numeric operator is device dependent. But the basic data type is always the same so the number of Modbus registers used is consistent. So the question is:
a) Do we still transmit the unscaled value and leave the interpretation (applying the numeric operator depending on the actual device) to the client? This might be OK because the client addresses the modbus server by the product ID so it knows exactly what to expect in principle.
b) apply the numeric operator in ems-esp and transmit the result, either as a IEEE 754 float or with a "worst case" numeric operator and possibly using more registers for the value
c) transmit the numeric operator in a separate register (this would practically double the number of registers used)

I think (a) is the option causing the least trouble.

**Reply by mheyse** [reply]
Hi, I just summited draft PR #1618 with a proof-of-concept implementation. I'd appreciate your comments to see if this is heading in the right direction. Thanks!

**Reply by proddy** [reply]
thanks, I'll take a peek at the PR. As for the `lib_deps` the reason for having locally stored libs as mainly because we're always tweaking some of them. The ones that are fixed (like ArduinoJson and your emodbus lib) and makes sense to have platformio manage them so absolutely fine if you want to add a lib_deps now.

**Reply by mheyse** [reply]
Regarding writing data through modbus. I guess the way to go is

- implement a `WRITE_HOLD_REGISTER` callback
- derive the entity shortname from the modbus register and construct a command path from this shortname (same as the mqtt topic)
- create an input Json Object with the value, hc etc.
- call `Command::process`

But then a lot of device specific stuff happens in the callback that's eventually executed. So this method does not at all work well with "raw" values (values that do not have the numeric_operator applied). So this method would really need the "final" values with all scale factors already applied.

So we're back at the problem of the data format. 

Of course read and write operations need to be symmetric, so if you read unscaled values you also need to write unscaled values. In this case the `WRITE_HOLD_REGISTER` callback could just apply the scale factor, resulting in a float value, and using that in the JsonObject for `Command::process`. But I'm a bit worried that we might lose precision with this approach. But if I understand the code correctly, mqtt does the same thing, right? Applying the factor and transmitting the value as a float. So this should be fine I guess?

**Comment by nanosonde** [comment]
Hi @mheyse !
I am interested in this combination EMS-ESP32<-ModbusTCP-Server->Weinzierl KNX Modbus TCP Gateway 716 secure.

Have you been able to finally use this combination in your production setup?

**Comment by mheyse** [comment]
Hi, yes, that's exactly what I'm doing in my home setup.

**Comment by nanosonde** [comment]
@mheyse 
Sounds great. Did you come across any issues or was it straight forward?
Do you have some config details on how you have setup things that are most important?

**Comment by mheyse** [comment]
Well, if you count implementing the Modbus connector for EMS-ESP as part of the setup, then no, it was not what I would consider straight forward :-)

Other than that it seems like everything is working as intended now. I'm using it for basic things like monitoring and most importantly for the summer-cooling/winter-heating setup which uses the EMS-ESP's simulated RC100H thermostat and feeds it via Modbus with temperature/humidity measurements that I get from KNX devices.

So if you follow the Modbus documentation, setup the KNX/Modbus gateway with the registers and scaling values listed in the table it should work right away.

FWIW, I'm still on an older dev firmware (3.7.0-dev.38), never bothered upgrading since all was working well.
